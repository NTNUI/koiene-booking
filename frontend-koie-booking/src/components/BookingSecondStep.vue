<template>
  <v-layout :class="$style.container">
    <v-layout :class="$style.separator">
      <h3 class="py-4" :class="$style.guestRegistration">{{ $t('booking.guest_registration') }}</h3>
      <v-form v-model="validForm" :class="$style.form">
        <v-layout v-for="i in Number(bedsChosen)" :key="i">
          <div :class="$style.fieldWrapper">
            <v-text-field
              v-if="i === 1"
              v-model="guests[i - 1].name"
              :label="$t('booking.full_name')"
              placeholder="Ola Nordmann"
              :rules="nameRules"
              :class="$style.nameField"
              required
              @blur="setGuests"
            />
            <v-text-field
              v-if="i !== 1"
              v-model="guests[i - 1].name"
              :label="$t('booking.full_name')"
              placeholder="Ola Nordmann"
              :rules="nameRules"
              :class="$style.nameField"
              required
              @blur="setGuests"
            />
            <v-text-field
              v-model="guests[i - 1].number"
              :label="$t('booking.number')"
              placeholder="95898788"
              number
              :rules="numberRules"
              :class="$style.numberField"
              required
              @blur="setGuests"
            />
            <div :class="$style.emailFieldWrapper">
              <v-text-field
                v-if="i === 1"
                v-model="guests[i - 1].email"
                :label="$vuetify.breakpoint.xs ? $t('booking.emailGeneric') : $t('booking.emailMain')"
                placeholder="ola.n@hotmail.com"
                :rules="emailStrictRules"
                required
              />
              <v-text-field
                v-if="i !== 1"
                v-model="guests[i - 1].email"
                :label="$vuetify.breakpoint.xs ? $t('booking.emailGeneric') : $t('booking.emailOptional')"
                placeholder="ola.n@hotmail.com"
                :rules="emailLooseRules"
              />
            </div>
            <v-checkbox
              v-model="guests[i - 1].isMember"
              class="mx-2"
              :label="$t('booking.member')"
              @change="setGuests"
            ></v-checkbox>
            <v-btn
              :disabled="i === 1"
              :color="$scssVars.globalColorBackgroundLight"
              :class="$style.rmvBtn"
              @click="removeGuestField(i)"
              >x</v-btn
            >
          </div>
        </v-layout>
        <v-btn :class="$style.addBtn" :color="$scssVars.globalColorPrimary" raised="true" @click="addGuestField"
          >+</v-btn
        >
      </v-form>
    </v-layout>
    <h3 :class="$style.price">Total price: {{ price }}kr</h3>
  </v-layout>
</template>

<script lang="ts">
import Vue from 'vue';
import { Guest, BookingSecondStepData } from '../types/booking';
export default Vue.extend({
  data(): BookingSecondStepData {
    return {
      bedsChosen: 1,
      bookingStep: 1,
      guests: [{ name: '', number: '', email: '', isMember: true, isMainBooker: true }],
      numberRules: [
        (e: any) => !!e || 'Number is required',
        (e: any) => e.length >= 8 || 'Number must be 8 characters'
      ],
      nameRules: [(e: any) => !!e || 'Name is required'],
      emailStrictRules: [
        (v: any) => !!v || 'E-mail is required',
        (v: any) => /.+@.+/.test(v) || 'E-mail must be valid'
      ],
      emailLooseRules: [(v: any) => v === '' || /.+@.+/.test(v) || 'E-mail must be valid'],

      edited: false,
      validForm: true,
      memberPrice: 0,
      nonMemberPrice: 0
    };
  },
  computed: {
    step(): number {
      return this.$store.state.booking.step;
    },
    guestField(): Array<Guest> {
      return this.guests;
    },
    beds(): number {
      return this.$store.state.booking.beds;
    },
    price(): number {
      let member = 0;
      let nonMember = 0;
      this.guests.forEach((i) => {
        if (i.isMember) {
          member++;
        } else {
          nonMember++;
        }
      });
      return this.memberPrice * member + this.nonMemberPrice * nonMember;
    }
  },
  watch: {
    validForm: function() {
      this.$store.dispatch('booking/SET_VALID_FORM', this.validForm);
    }
  },

  mounted() {
    this.edited = this.$store.state.booking.edited;
    this.bedsChosen = Number(this.$store.state.booking.beds);
    if (this.edited === true) {
      this.guests = this.$store.state.booking.guests;
    } else {
      this.addGuests();
    }
    this.bookingStep = Number(this.$store.state.booking.step);
    this.$store.dispatch('booking/SET_VALID_FORM', this.validForm);
    this.memberPrice = this.$store.state.koie.koieData.price_member;
    this.nonMemberPrice = this.$store.state.koie.koieData.price_not_member;
  },
  methods: {
    setGuests() {
      this.$store.dispatch('booking/SET_GUESTS', this.guests);
      this.$store.dispatch('booking/SET_EDITED', true);
    },
    addGuests() {
      for (let i = 0; i < this.bedsChosen - 1; i++) {
        this.guests.push({ name: '', number: '', isMember: true, isMainBooker: false });
      }
    },
    addGuestField() {
      if (this.bedsChosen < this.$store.state.booking.availableBeds) {
        this.guests.push({ name: '', number: '', isMember: true, isMainBooker: false });
        this.bedsChosen += 1;
        this.$store.dispatch('booking/SET_BEDS', this.bedsChosen);
      }
    },
    removeGuestField(index: number) {
      if (this.guests.length > 1) {
        this.guests.splice(index - 1, 1);
        this.bedsChosen -= 1;
        this.$store.dispatch('booking/SET_GUESTS', this.guests);
        this.$store.dispatch('booking/SET_BEDS', this.bedsChosen);
      }
    }
  }
});
</script>

<style lang="scss" module>
.separator {
  border-left: solid 4px;
  border-color: #2f3a50;
  display: flex;
  flex-direction: column;
  margin-top: 24px;
}
.container {
  display: flex;
  flex-direction: column;
}

.form,
.guestRegistration {
  padding-left: 16px;
  padding-right: 16px;
  @media only screen and (max-width: 600px) {
    padding-left: 8px;
    padding-right: 8px;
  }
}

.fieldWrapper {
  display: flex;
  width: 100%;
  align-items: center;
}

.emailFieldWrapper {
  width: 50%;
}

.nameField {
  width: 40%;
  padding-right: 10px;
}

.numberField {
  padding-right: 10px;
}
.rmvBtn {
  margin-left: 20px;
  @media only screen and (max-width: 600px) {
    margin-left: 8px;
    padding-left: 0px !important;
    padding-right: 0px !important;
    min-width: 30px !important;
  }
}
.price {
  text-align: center;
  margin-top: 30px;
}

.addBtn {
  margin-top: 16px;
}
</style>
