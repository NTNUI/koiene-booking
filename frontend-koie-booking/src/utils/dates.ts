import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
