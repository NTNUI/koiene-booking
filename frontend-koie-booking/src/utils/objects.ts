export const clone = (object: any) => Object.assign({}, object);

export const renameKey = (object: any, key: string, newKey: string): any => {
  const value = object[key];
  delete object[key];
  object[newKey] = value;
  return object;
};

export const sortObjectByKey = (unordered: any) => {
  let ordered: any;
  ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key) {
      ordered[key] = unordered[key];
    });
  return ordered;
};
