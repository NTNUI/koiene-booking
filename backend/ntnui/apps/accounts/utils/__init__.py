from accounts.utils.mail_utils import send_gsuite_confirmation_mail
