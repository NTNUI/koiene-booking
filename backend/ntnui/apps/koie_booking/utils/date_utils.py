from django.utils.timezone import timedelta


def get_daterange(start_date, end_date):
    for n in range(int((end_date - start_date).days)):
        yield start_date + timedelta(n)
