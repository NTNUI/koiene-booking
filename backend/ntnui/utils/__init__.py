from ntnui.utils.pagination import PageSize7, PageSize20, PageSize100
