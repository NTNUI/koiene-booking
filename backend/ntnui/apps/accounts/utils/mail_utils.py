from django.core.mail import send_mail
from django.template.loader import render_to_string

from ntnui.enums import MembershipType


def new_membership(membership):
    """ Sends confirmation email to users and guests who signs up for events and sub-events. """

    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [membership.member.email]
    subject = "Welcome"

    context = {"user": membership.member, "group": membership.group}

    # Mail body.
    plain_message = render_to_string("emails/new_membership.txt", context)
    html_message = render_to_string("emails/new_membership.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_gsuite_confirmation_mail(user, contact_email, password):
    # Mail header.
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = "Opprettelse av ny NTNUI-bruker"

    context = {
        "first_name": user.first_name,
        "email": user.email,
        "contact_email": contact_email,
        "password": password,
    }

    # Mail body.
    plain_message = render_to_string("emails/new_gsuite_account.txt", context)
    html_message = render_to_string("emails/new_gsuite_account.html", context)

    send_mail(subject, plain_message, sender, receiver, html_message=html_message)


def send_welcome_mail_to_leader(user, group):
    """ Send a mail to a new leader """
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny leder i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_leader_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_leader_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_cashier(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny kasserer i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_cashier_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_cashier_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail_to_board_member(user, group):
    """Send a mail to the new cashier"""
    sender = "noreply@mg.ntnui.no"
    receiver = [user.email]
    subject = f"Velkommen som ny tillitsvalgt i {group}"

    context = {
        "first_name": user.first_name,
        "email": user.get_contact_email(),
        "group": group.name,
    }

    plain_message = render_to_string(
        template_name="emails/new_board_membership.txt", context=context
    )
    html_message = render_to_string(
        template_name="emails/new_board_membership.html", context=context
    )

    send_mail(
        subject=subject,
        message=plain_message,
        from_email=sender,
        recipient_list=receiver,
        html_message=html_message,
    )


def send_welcome_mail(membership):
    """ Delegate welcome mail to different board members """
    send_welcome_mail_to_board_member(user=membership.member, group=membership.group)
    if membership.type == MembershipType.leader:
        send_welcome_mail_to_leader(user=membership.member, group=membership.group)

    if membership.type == MembershipType.cashier:
        send_welcome_mail_to_cashier(user=membership.member, group=membership.group)
