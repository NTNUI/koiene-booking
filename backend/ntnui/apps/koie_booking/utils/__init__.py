from koie_booking.utils.mail_utils import send_confirmation_mail, send_koie_information_mail
from koie_booking.utils.date_utils import get_daterange
