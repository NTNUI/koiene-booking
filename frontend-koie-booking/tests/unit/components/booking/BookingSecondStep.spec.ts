import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.use(Vuetify);

// Components or views
import BookingSecondStep from '@/components/booking/BookingSecondStep.vue';
import { createWrapper } from '../../utils';

describe('Component BookingSecondStep.vue', () => {
  it('Matches snapshot', async () => {
    const wrapper = createWrapper(BookingSecondStep, {
      sync: false,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Calculate total price of 1x member, 2x non-member', async () => {
    const wrapper: any = createWrapper(BookingSecondStep, {
      data() {
        return {
          guests: [
            { name: '', number: '', email: '', isMember: true },
            { name: '', number: '', isMember: false },
            { name: '', number: '', isMember: false },
          ],
        };
      },
      sync: false,
    });

    expect(wrapper.vm.price).toBe(200);
  });
});
