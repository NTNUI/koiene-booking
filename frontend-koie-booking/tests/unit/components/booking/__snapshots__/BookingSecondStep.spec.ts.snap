// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component BookingSecondStep.vue Matches snapshot 1`] = `
<div class="layout container">
  <div class="layout separator">
    <h3 class="py-4 guestRegistration">Guest Registration</h3>
    <form novalidate="novalidate" class="v-form form">
      <div class="layout">
        <div class="fieldWrapper">
          <div class="v-input theme--light v-text-field v-text-field--placeholder nameField">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-8" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Full name</label><input required="required" id="input-8" placeholder="Ola Nordmann" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <!---->
          <div class="v-input theme--light v-text-field v-text-field--placeholder numberField">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-11" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Phone number</label><input number="" required="required" id="input-11" placeholder="95898788" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
          <div class="emailFieldWrapper">
            <div class="v-input theme--light v-text-field v-text-field--placeholder">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-14" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">E-mail (main contact person)</label><input required="required" id="input-14" placeholder="ola.n@hotmail.com" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input mx-2 v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--checkbox primary--text">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-input--selection-controls__input"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-marked theme--light primary--text"></i><input aria-checked="true" id="input-17" role="checkbox" type="checkbox" value="">
                  <div class="v-input--selection-controls__ripple primary--text"></div>
                </div><label for="input-17" class="v-label theme--light" style="left: 0px; position: relative;">NTNUI Member</label>
              </div>
              <div class="v-messages theme--light primary--text"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div> <button type="button" disabled="disabled" class="v-btn v-btn--contained v-btn--disabled theme--light v-size--default rmvBtn"><span class="v-btn__content">x</span></button>
        </div>
      </div> <button type="button" class="v-btn v-btn--contained theme--light v-size--default addBtn" raised="true"><span class="v-btn__content">+</span></button>
    </form>
  </div>
  <h3 class="price">Total price: 0kr</h3>
</div>
`;
